File: app/src/main/java/com/qimian233/ztool/hook/modules/systemui/NotificationIconHook.java
package com.qimian233.ztool.hook.modules.systemui;

import android.annotation.SuppressLint;
import android.content.SharedPreferences;

import com.qimian233.ztool.hook.base.BaseHookModule;

import io.github.libxposed.api.XposedModuleInterface;

import java.lang.reflect.Constructor;
import java.lang.reflect.Field;

/**
 * SystemUI通知图标限制Hook模块
 * 功能：修改状态栏通知图标的最大显示数量限制
 * 支持Android 12+的SystemUI架构
 */
@SuppressLint("PrivateApi")
public class NotificationIconHook extends BaseHookModule {

    private int NEW_MAX_ICONS;
    private static final String PREFS_NAME = "StatusBar_notifyNumSize";

    public NotificationIconHook() {}

    @Override
    public String getModuleName() {
        return "notification_icon_limit";
    }

    @Override
    public String[] getTargetPackages() {
        return new String[]{
                "com.android.systemui"
        };
    }

    @Override
    public void handleLoadPackage(XposedModuleInterface.PackageLoadedParam param) throws Throwable {
        ClassLoader classLoader = param.getDefaultClassLoader();
        String packageName = param.getPackageName();
        if ("com.android.systemui".equals(packageName)) {
            SharedPreferences prefs = this.xposed.getRemotePreferences(PREFS_NAME);
            NEW_MAX_ICONS = prefs.getInt("notify_num_size", 4);
            hookSystemUIIconLimit(classLoader);
        }
    }

    private void hookSystemUIIconLimit(ClassLoader classLoader) {
        log("开始 Hook SystemUI 通知图标限制，设置最大图标数: " + NEW_MAX_ICONS);

        try {
            // Hook 1: 修改资源获取的最大图标数量
//            hookResourceInteger(classLoader);

            // Hook 2: 修改 NotificationIconContainerStatusBarViewModel 的 maxIcons 字段
            hookViewModelConstructor(classLoader);

            // Hook 3: 修改 NotificationIconsViewData 构造函数，应用数量限制
            hookViewDataConstructor(classLoader);

            log("SystemUI 通知图标限制Hook设置完成");
        } catch (Throwable e) {
            logError("SystemUI Hook过程中发生错误", e);
        }
    }

    private void hookViewModelConstructor(ClassLoader classLoader) {
        try {
            Class<?> viewModelClass = classLoader.loadClass(
                    "com.android.systemui.statusbar.notification.icon.ui.viewmodel.NotificationIconContainerStatusBarViewModel"
            );

            Constructor<?> ctor = viewModelClass.getDeclaredConstructor(
                    classLoader.loadClass("kotlin.coroutines.CoroutineContext"),
                    classLoader.loadClass("com.android.systemui.statusbar.phone.domain.interactor.DarkIconInteractor"),
                    classLoader.loadClass("com.android.systemui.statusbar.notification.icon.domain.interactor.StatusBarNotificationIconsInteractor"),
                    classLoader.loadClass("com.android.systemui.statusbar.notification.domain.interactor.HeadsUpNotificationIconInteractor"),
                    classLoader.loadClass("com.android.systemui.keyguard.domain.interactor.KeyguardInteractor"),
                    android.content.res.Resources.class,
                    classLoader.loadClass("com.android.systemui.shade.domain.interactor.ShadeInteractor")
            );

            hookWithId(ctor, "ctor_1", chain -> {
                // after constructor: chain.proceed() then set field
                chain.proceed();
                try {
                    Field myField = chain.getThisObject().getClass().getDeclaredField("maxIcons");
                    myField.setAccessible(true);
                    myField.setInt(chain.getThisObject(), NEW_MAX_ICONS);
                    log("成功修改 ViewModel maxIcons 为 " + NEW_MAX_ICONS);
                } catch (Exception e) {
                    logError("修改 ViewModel maxIcons 字段失败", e);
                }
                return null;
            });

            log("ViewModel构造函数Hook设置成功");
        } catch (Throwable e) {
            log("找不到 ViewModel 类，可能系统版本不兼容: " + e.getMessage());
        }
    }

    private void hookViewDataConstructor(ClassLoader classLoader) {
        try {
            Class<?> viewDataClass = classLoader.loadClass(
                    "com.android.systemui.statusbar.notification.icon.ui.viewmodel.NotificationIconsViewData"
            );

            Constructor<?> ctor = viewDataClass.getDeclaredConstructor(
                    java.util.List.class,
                    int.class,
                    classLoader.loadClass("com.android.systemui.statusbar.notification.icon.ui.viewmodel.NotificationIconsViewData$LimitType")
            );

            hookWithId(ctor, "ctor_2", chain -> {
                try {
                    // 获取图标列表
                    Object iconList = chain.getArg(0);
                    int listSize = getListSize(iconList);

                    // 使用NEW_MAX_ICONS作为限制，但不超过实际图标数量
                    int effectiveLimit = Math.min(NEW_MAX_ICONS, listSize);
                    int currentLimit = (int) chain.getArg(1);

                    // 只有当当前限制不等于我们设置的有效限制时才修改
                    if (currentLimit != effectiveLimit) {
                        log("修改图标限制 " + currentLimit + " -> " + effectiveLimit + " (图标总数: " + listSize + ")");
                        return chain.proceed(new Object[]{iconList, effectiveLimit, chain.getArg(2)});
                    }
                } catch (Exception e) {
                    logError("ViewData Hook过程中发生错误", e);
                }
                return chain.proceed();
            });

            log("ViewData构造函数Hook设置成功");
        } catch (Throwable e) {
            log("找不到 ViewData 类，可能系统版本不兼容: " + e.getMessage());
        }
    }

    // 辅助方法：获取列表大小
    private int getListSize(Object list) {
        try {
            return (int) list.getClass().getDeclaredMethod("size").invoke(list);
        } catch (Exception e) {
            return 0;
        }
    }
}

File: app/src/main/java/com/qimian233/ztool/hook/modules/systemui/statusbar/NotificationIconHook.kt
package com.qimian233.ztool.hook.modules.systemui.statusbar

import android.annotation.SuppressLint
import android.content.SharedPreferences
import com.qimian233.ztool.data.keys.PreferenceKeys
import com.qimian233.ztool.data.keys.ScopeKeys
import com.qimian233.ztool.hook.base.AppHookModule
import io.github.libxposed.api.XposedModuleInterface.PackageLoadedParam
import java.lang.reflect.Constructor
import java.lang.reflect.Field

/**
 * SystemUI notification icon limit hook module.
 * Function: modifies the maximum display count limit for status bar notification icons.
 * Supports the Android 12+ SystemUI architecture.
 */
@SuppressLint("PrivateApi")
class NotificationIconHook : AppHookModule() {

    private var newMaxIcons = 0

    override fun getModuleName(): String = PreferenceKeys.NOTIFICATION_ICON_LIMIT.name

    override fun getTargetPackages(): Array<String> = arrayOf(ScopeKeys.SYSTEM_UI.packageName)

    override fun handleLoadPackage(param: PackageLoadedParam) {
        val classLoader = param.defaultClassLoader
        val packageName = param.packageName
        if (ScopeKeys.SYSTEM_UI.packageName == packageName) {
            val prefs: SharedPreferences = xposed.getRemotePreferences(PREFS_NAME)
            newMaxIcons = prefs.getInt("notify_num_size", 4)
            hookSystemUIIconLimit(classLoader)
        }
    }

    private fun hookSystemUIIconLimit(classLoader: ClassLoader) {
        logger.info("Hooking SystemUI notification icon limit, max icons: $newMaxIcons")

        try {
            // Hook 1: modify the max icon count obtained from resources
//            hookResourceInteger(classLoader);

            // Hook 2: modify the maxIcons field of NotificationIconContainerStatusBarViewModel
            hookViewModelConstructor(classLoader)

            // Hook 3: modify the NotificationIconsViewData constructor to apply the count limit
            hookViewDataConstructor(classLoader)

            logger.info("SystemUI notification icon limit hooks installed")
        } catch (e: Throwable) {
            logger.error("Error during SystemUI hook", e)
        }
    }

    private fun hookViewModelConstructor(classLoader: ClassLoader) {
        try {
            val viewModelClass = classLoader.loadClass(
                "com.android.systemui.statusbar.notification.icon.ui.viewmodel.NotificationIconContainerStatusBarViewModel"
            )

            val ctor: Constructor<*> = viewModelClass.getDeclaredConstructor(
                classLoader.loadClass("kotlin.coroutines.CoroutineContext"),
                classLoader.loadClass("com.android.systemui.statusbar.phone.domain.interactor.DarkIconInteractor"),
                classLoader.loadClass("com.android.systemui.statusbar.notification.icon.domain.interactor.StatusBarNotificationIconsInteractor"),
                classLoader.loadClass("com.android.systemui.statusbar.notification.domain.interactor.HeadsUpNotificationIconInteractor"),
                classLoader.loadClass("com.android.systemui.keyguard.domain.interactor.KeyguardInteractor"),
                android.content.res.Resources::class.java,
                classLoader.loadClass("com.android.systemui.shade.domain.interactor.ShadeInteractor")
            )

            hookWithId(ctor, "ctor_1") { chain ->
                // after constructor: chain.proceed() then set field
                chain.proceed()
                try {
                    val myField: Field = chain.thisObject.javaClass.getDeclaredField("maxIcons")
                    myField.isAccessible = true
                    myField.setInt(chain.thisObject, newMaxIcons)
                    logger.debug("Successfully set ViewModel maxIcons to $newMaxIcons")
                } catch (e: Exception) {
                    logger.error("Failed to modify ViewModel maxIcons field", e)
                }
                null
            }

            logger.info("ViewModel constructor hook applied")
        } catch (e: Throwable) {
            logger.warn("ViewModel class not found, system version may be incompatible: " + e.message)
        }
    }

    private fun hookViewDataConstructor(classLoader: ClassLoader) {
        try {
            val viewDataClass = classLoader.loadClass(
                "com.android.systemui.statusbar.notification.icon.ui.viewmodel.NotificationIconsViewData"
            )

            val ctor: Constructor<*> = viewDataClass.getDeclaredConstructor(
                List::class.java,
                Int::class.javaPrimitiveType,
                classLoader.loadClass("com.android.systemui.statusbar.notification.icon.ui.viewmodel.NotificationIconsViewData\$LimitType")
            )

            hookWithId(ctor, "ctor_2") { chain ->
                try {
                    // Get the icon list
                    val iconList = chain.args[0]
                    val listSize = getListSize(iconList)

                    // Use NEW_MAX_ICONS as the limit, but not exceeding the actual icon count
                    val effectiveLimit = minOf(newMaxIcons, listSize)
                    val currentLimit = chain.args[1] as Int

                    // Only modify when the current limit differs from our effective limit
                    if (currentLimit != effectiveLimit) {
                        logger.debug("Icon limit changed $currentLimit -> $effectiveLimit (total icons: $listSize)")
                        return@hookWithId chain.proceed(arrayOf(iconList, effectiveLimit, chain.args[2]))
                    }
                } catch (e: Exception) {
                    logger.error("Error during ViewData hook", e)
                }
                chain.proceed()
            }

            logger.info("ViewData constructor hook applied")
        } catch (e: Throwable) {
            logger.warn("ViewData class not found, system version may be incompatible: " + e.message)
        }
    }

    // Helper: get the list size
    private fun getListSize(list: Any?): Int {
        return try {
            list!!.javaClass.getDeclaredMethod("size").invoke(list) as Int
        } catch (e: Exception) {
            0
        }
    }

    companion object {
        private const val PREFS_NAME = "StatusBar_notifyNumSize"
    }
}
